Updating PyXet Copy logic

hoytak opened the pull request:
Updates pyxet copy logic to make it more readable.
Needed for progress bar work.

seanses commented on python/pyxet/pyxet/file_operations.py:
```diff
@@ -0,0 +1,329 @@
+
+        # Figure out what the destination is like, etc. 
+        try:
+            info = dest_fs.info(dest_path)
```
Since info is only used to check whether the path is directory-like, we can just reuse the `_isdir` function

seanses commented on python/pyxet/pyxet/file_operations.py:
```diff
@@ -0,0 +1,329 @@
+    _single_file_copy_impl(cp_unit, src_fs, dest_fs)
+
+
+def _build_cp_action_list_impl(src_fs, src_path, dest_fs, dest_path, recursive):
```
This function contains long nested if-else blocks and there are many places checking `is_dir` so it's a bit difficult to follow. Is there a reason to not use the previous version?

hoytak replied:
I found the other one difficult to follow, so I laid out all the different cases as a tree of if statements, since they had slightly different behaviors.  I can definitely comment it better to make it more readable what the different cases are.

seanses commented on python/pyxet/pyxet/file_operations.py:
```diff
@@ -0,0 +1,370 @@
+            src_path = src_path.rstrip("/")
+            src_info = src_fs.info(src_path)
+            src_is_directory = True
+            if not src_info['type'] in ["directory", "branch"]: 
```
Let's update all these manual checks to call `_isdir`

hoytak replied:
So the issue is that _isdir performs a call to fs.info, which is a remote call to the server if it's a remote file system.  So ideally we should limit calls to that if we can get it another way.  Here we need the size, so making one call to fs.info for that, or obtaining it from the find method that returns it in bulk, would be way faster.

seanses replied:
Good point

seanses commented:
I think the failed test test_glob_recursive_upload on Windows is due to path name `subdir\\data` not correctly translated to `subdir/data`

hoytak commented:
> I think the failed test test_glob_recursive_upload on Windows is due to path name `subdir\data` not correctly translated to `subdir/data`. The existing code explicitly replaces `\` with `/` : https://github.com/xetdata/pyxet/blob/0add233cfb1d56b4fbcb45dd7198c565f37d3cb9/python/pyxet/pyxet/util.py#L200C17-L200C17

Thank you!  Fixed.  Added in to the path_normalize routine -- which is what should be anyway...

seanses commented on python/pyxet/pyxet/util.py:
```diff
@@ -12,6 +12,16 @@
 MAX_CONCURRENT_COPIES = threading.Semaphore(32)
 CHUNK_SIZE = 16 * 1024 * 1024
 
+class CopyUnit:
```
redundant? there is one in the file_operations.py

hoytak replied:
Good catch.

seanses commented on python/pyxet/pyxet/util.py:
```diff
@@ -33,11 +43,10 @@ def _should_load_aws_credentials():
 
 
-def _get_fs_and_path(uri):
+def _get_fs_and_path(uri, strip_trailing_slash = True):
```
this param `strip_trailing_slash` doesn't seem to be used anywhere?

hoytak replied:
Very good catch.  It should be passed to _path_normalize.

hoytak replied:
Very good catch.  It should be passed to _path_normalize.

seanses commented on python/pyxet/pyxet/file_operations.py:
```diff
@@ -0,0 +1,369 @@
+    
+    
+    
+    # This function has two parts.  First, we we set a number of variables that then determine how 
```
typo, extra `we`

seanses reviewed (approved):
Looks good! Some nit comments above
